update-formula.1
.TH update-formula "1" "December 15, 2023" "" "update-formula manual"

.SH NAME
update-formula - updates assets referred to in Homebrew formula files

.SH SYNOPSIS
.B update-formula
[-d|--dry-run] [-V|--version] [-h|--help]
[
.I file ...
]
.RE

.SH DESCRIPTION
Updates version numbers and SHA-256 hashes in Homebrew formula files for GitHub
releases. Helps updating a tap after publishing a new release on GitHub with new
binary assets to install using Homebrew.

.SH ARGUMENTS
.B
.IP "file"
formula file to update

.SH OPTIONS
.B
.IP "-d|--dry-run"
only print what would be done without doing it
.B
.IP "-V, --version"
print the version of the executable and exit
.B
.IP "-h, --help"
print the usage information and exit

.RE
If no input file is specified, it will be read from standard input.
If multiple files are specified and file overwriting is not enabled,
the files will be only checked and their names printed out.

.SH EXAMPLES
.B
.IP "update-formula yaml2json.rb"
Update a formula to the latest version released on GitHub.

.SH BUGS
update-formula's bug tracker is located at <https://github.com/prantlf/v-update-formula/issues>

.SH DOCUMENTATION
See the website for documentation <https://github.com/prantlf/v-update-formula#readme>

.SH AUTHOR
Currently maintained by Ferdinand Prantl <prantlf@gmail.com>


CHANGELOG.md
# Changes

## [0.1.4](https://github.com/prantlf/update-formula/compare/v0.1.3...v0.1.4) (2023-12-17)

### Bug Fixes

* Close file before logging to get correct timing ([e8b1344](https://github.com/prantlf/update-formula/commit/e8b13446e321328554d6f0859966a1525e016943))

## [0.1.3](https://github.com/prantlf/update-formula/compare/v0.1.2...v0.1.3) (2023-12-17)

### Bug Fixes

* Reuse code from prantlf.github ([ef1d1eb](https://github.com/prantlf/update-formula/commit/ef1d1eb37d770db3cc998bed21ddb26cf54286b6))

## [0.1.2](https://github.com/prantlf/update-formula/compare/v0.1.1...v0.1.2) (2023-12-15)

### Bug Fixes

* Include man pages in asset archives ([f8fcf57](https://github.com/prantlf/update-formula/commit/f8fcf572418d265e5224e5fe2c739ea7a8468df8))

## [0.1.1](https://github.com/prantlf/update-formula/compare/v0.1.0...v0.1.1) (2023-12-15)

### Bug Fixes

* Add whitespace to logging ([e8d2cc8](https://github.com/prantlf/update-formula/commit/e8d2cc875d60b29a54f1a3884b7b73ef5b1d18ac))

## [0.1.0](https://github.com/prantlf/update-formula/compare/v0.0.0...v0.1.0) (2023-12-15)

### Features

* Initial release ([399ac57](https://github.com/prantlf/update-formula/commit/399ac57bb8dcda1a3658b4ef98427426bed77359))

## 0.0.0 (2023-12-15)

Project creation.


src/update-formula.v
import crypto.sha256 { hexhash }
import os { create, read_lines }
import prantlf.cli { Cli, Env, run }
import prantlf.debug { new_debug }
import prantlf.github { download_asset, get_gh_token, get_latest_release }
import prantlf.pcre { pcre_compile }

const d = new_debug('updtap')

const version = '0.1.4'

const usage = 'Updates version numbers and SHA-256 hashes in Homebrew formula files for GitHub releases.

Usage: update-formula [options] [<file> ...]

  <file>        formula file to update

Options:
  -d|--dry-run  only print what would be done without doing it
  -V|--version  print the version of the executable and exit
  -h|--help     print the usage information and exit

Examples:
  $ update-formula yaml2json.rb'

struct Opts {
	dry_run bool
mut:
	gh_token string
}

const re_home = pcre_compile('homepage "https://github.com/(.+)"', 0) or { panic('re_homepage') }
const re_ver = pcre_compile('version "([.0-9]+)"', 0) or { panic('re_version') }
const re_url = pcre_compile('url "https://github.com/.+/releases/download/v([^/]+)/(.+)"',
	0) or { panic('re_url') }
const re_hash = pcre_compile('sha256 "(.+)"', 0) or { panic('re_hash') }
const re_tag = pcre_compile('"tag_name"\\s*:\\s*"v([.0-9]+)"', 0) or { panic('re_tag') }

fn get_latest(repo string, token string) !string {
	body := get_latest_release(repo, token)!
	if body.len == 0 {
		return error('no release found in ${repo}')
	}
	if m := re_tag.exec(body, 0) {
		return m.group_text(body, 1) or { panic('') }
	}
	return error('no tag found in ${repo}/releases/latest')
}

fn get_hash(url string, token string) !string {
	body := download_asset(url, token)!
	return hexhash(body)
}

fn update(file string, opts &Opts) ! {
	println('file: ${file}')

	d.log('reading "%s"', file)
	mut lines := read_lines(file)!

	d.log_str('parsing contents')
	mut repo := ''
	mut new_ver := ''
	mut url := ''
	for i, line in lines {
		if m := re_home.exec(line, 0) {
			repo = m.group_text(line, 1) or { panic('') }
			println('  repository: "${repo}"')
		} else if m := re_ver.exec(line, 0) {
			if repo.len == 0 {
				return error('version spotted before homepage')
			}
			ver := m.group_text(line, 1) or { panic('') }
			new_ver = get_latest(repo, opts.gh_token)!
			if ver == new_ver {
				println('  version ${ver} is already the latest')
				return
			}
			println('  updating version ${ver} to ${new_ver}')
			start, end := m.group_bounds(1) or { panic('') }
			lines[i] = line[..start] + new_ver + line[end..]
		} else if m := re_url.exec(line, 0) {
			if new_ver.len == 0 {
				return error('url spotted before version')
			}
			asset := m.group_text(line, 2) or { panic('') }
			println('  asset: "${asset}"')
			url = 'https://github.com/${repo}/releases/download/v${new_ver}/${asset}'
			start, end := m.group_bounds(1) or { panic('') }
			lines[i] = line[..start] + new_ver + line[end..]
			continue
		} else if m := re_hash.exec(line, 0) {
			if url.len == 0 {
				return error('hash not preceded with url')
			}
			hash := get_hash(url, opts.gh_token)!
			println('  hash: "${hash}"')
			start, end := m.group_bounds(1) or { panic('') }
			lines[i] = line[..start] + hash + line[end..]
		}
		url = ''
	}

	if opts.dry_run {
		d.log_str('  not writing (dry-run)')
		return
	}

	d.log_str('reopening the file')
	mut out := create(file)!
	defer {
		out.close()
	}

	d.log_str('writing new contents')
	for line in lines {
		out.writeln(line)!
	}
	out.close()
	d.log_str('file written')
}

fn body(mut opts Opts, args []string) ! {
	if args.len == 0 {
		return error('missing file')
	}

	if opts.gh_token.len == 0 {
		opts.gh_token = get_gh_token()!
	}

	for arg in args {
		update(arg, opts)!
	}
}

fn main() {
	run(Cli{
		usage: usage
		version: version
		env: Env.both
	}, body)
}
